File: scripts/check_docstrings.py
"""Check that all code blocks in docstrings are properly closed."""

import ast
import re
from enum import StrEnum
from pathlib import Path
from typing import NamedTuple, TypeIs

from rich.console import Console
from rich.table import Table
from rich.text import Text

from pyochain import (
    NONE,
    Err,
    Iter,
    Null,
    Ok,
    Option,
    Result,
    Seq,
    SetMut,
    Some,
    Vec,
    option,
)
from pyochain.abc import PyoIterator

CODE_BLOCK_PATTERN = re.compile(r"^```(\w*)", re.MULTILINE)
SKIP_DECORATORS = SetMut.from_ref({"overload", "override", "wraps", "property"})
type MethodNode = ast.FunctionDef | ast.AsyncFunctionDef
type DocumentableNode = MethodNode | ast.ClassDef


ROOT = Path()
SRC_DIR = ROOT.joinpath("pyochain")
CONSOLE = Console()


class Color(StrEnum):
    """Enum for consistent console message styling."""

    SUCCESS = "green"
    INFO = "cyan bold"
    WARNING = "yellow"
    ERROR = "red bold"
    BLANK = "white"

    def show(self, msg: str) -> None:
        """Print a message with the color."""
        CONSOLE.print(Text(msg, style=self.value))


class DocstringInfo(NamedTuple):
    """Information about a docstring."""

    name: str
    line_no: int
    content: str


class ErrorKind(StrEnum):
    """Kind of error found in a docstring."""

    MISSING = "Missing docstring"
    NO_DOCTEST = "Missing doctest: No ```python block found in docstring"
    UNCLOSED = "Unclosed ``` block"
    CLOSED_NOT_OPENED = "Closing ``` block without opening"


class DocstringError(NamedTuple):
    """Error found in a docstring."""

    file_path: Path
    func_name: str
    line_no: int
    error_line_no: int
    errors: Seq[ErrorKind]


class ErrorDetail(NamedTuple):
    """Detail of an error with its line number."""

    line_no: int
    message: ErrorKind


class State(NamedTuple):
    """State during code block traversal."""

    errors: Vec[ErrorDetail]
    stack: Vec[int]


def main() -> None:
    """Check all docstrings in the project."""
    msg = "Checking docstrings for properly closed code blocks..."
    Color.INFO.show(msg)

    return (
        _get_files("pyi")
        .iter()
        .filter_false(lambda f: f.name == "_types.pyi")
        .flat_map(_check_file)
        .collect(Seq)
        .then(_handle_errors)
        .unwrap_or_else(lambda: Color.SUCCESS.show("[OK] No issues found!"))
    )


def _get_files(pattern: str) -> Seq[Path]:
    files = Seq(SRC_DIR.rglob(f"*.{pattern}"))
    Color.INFO.show(f"Checking {files.len()} {pattern} files...")
    return files


def _handle_errors(all_errors: Seq[DocstringError]) -> None:
    _show_table(all_errors)
    msg = f"[FAILED] Found {all_errors.len()} issue(s)"
    Color.ERROR.show(msg)


def _show_table(all_errors: Seq[DocstringError]) -> None:
    def _add_row(error: DocstringError) -> None:
        file = f"{error.file_path.relative_to(ROOT)}:{error.error_line_no}"
        table.add_row(
            file, error.func_name, error.errors.iter().map(lambda e: e.value).join("\n")
        )

    table: Table = Table(title="Issues Found", show_header=True)
    table.add_column("File", style="cyan")
    table.add_column("Function", style="magenta")
    table.add_column("Error", style="red")
    all_errors.iter().for_each(_add_row)
    CONSOLE.print(table)


def _check_file(file_path: Path) -> PyoIterator[DocstringError]:

    def _is_documentable(node: ast.AST) -> TypeIs[DocumentableNode]:
        return isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef, ast.ClassDef))

    tree = ast.parse(file_path.read_text(encoding="utf-8"))

    return (
        Iter(ast.walk(tree))
        .filter(_is_documentable)  # We do two filter for `TypeIs` inference
        .filter(lambda node: not _has_skip_decorator(node))
        .filter_map(lambda node: _process_node(file_path, node))
    )


def _process_node(file_path: Path, node: DocumentableNode) -> Option[DocstringError]:
    match option(ast.get_docstring(node)):
        case Null() if _should_report_missing_docstring(node):
            return Some(
                DocstringError(
                    file_path,
                    node.name,
                    node.lineno,
                    node.lineno,
                    Seq([ErrorKind.MISSING]),
                )
            )
        case Some(docstring):
            result = _check_code_blocks(
                docstring,
                _get_docstring_start_line(node),
                skip_doctest=_has_skip_decorator(node),
            )
            match result:
                case Err(errors):
                    err = DocstringError(
                        file_path,
                        node.name,
                        node.lineno,
                        errors.first().line_no,
                        errors.iter().map(lambda e: e.message).collect(Seq),
                    )
                    return Some(err)
                case Ok(_):
                    return NONE
        case _:
            return NONE


def _should_report_missing_docstring(node: DocumentableNode) -> bool:
    return isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)) and (
        not _has_skip_decorator(node)
    )


def _check_code_blocks(
    docstring: str, start_line: int, *, skip_doctest: bool = False
) -> Result[tuple[()], Vec[ErrorDetail]]:
    def _process_line(state: State, line_num: int, line: str) -> State:
        marker = "```"
        stripped_line = line.lstrip()
        match = CODE_BLOCK_PATTERN.search(stripped_line)
        if not (match and stripped_line.startswith(marker)):
            return state
        if stripped_line == marker:
            if not state.stack.is_empty():
                _ = state.stack.pop()
                return state
            state.errors.append(
                ErrorDetail(start_line + line_num, ErrorKind.CLOSED_NOT_OPENED)
            )
            return state
        state.stack.append(line_num + 1)
        return state

    lines = Vec.wrap(docstring.split("\n"))
    state = State(Vec(), Vec())
    block_errors = (
        lines
        .iter()
        .enumerate()
        .fold_star(state, _process_line)
        .errors.iter()
        .chain(
            state.stack.iter().map(
                lambda idx: ErrorDetail(start_line + idx - 1, ErrorKind.UNCLOSED)
            )
        )
        .collect(Vec)
    )
    doctest_errors = _check_errs(lines, start_line, skip_doctest=skip_doctest)

    match doctest_errors:
        case Ok(_) if block_errors.is_empty():
            return Ok(())
        case Ok(_):
            return Err(block_errors)
        case Err(errors):
            errors.extend(block_errors)
            return Err(errors)


def _get_docstring_start_line(node: DocumentableNode) -> int:
    match node.body:
        case [ast.Expr(value=ast.Constant(value=str())), *_] as body:
            _ = body
            return node.body[0].lineno
        case _:
            return node.lineno


def _has_skip_decorator(node: DocumentableNode) -> bool:
    return Iter(node.decorator_list).any(
        lambda d: (
            (isinstance(d, ast.Name) and d.id in SKIP_DECORATORS)
            or (isinstance(d, ast.Attribute) and d.attr in SKIP_DECORATORS)
        )
    )


def _check_errs(
    lines: Vec[str], start_line: int, *, skip_doctest: bool
) -> Result[tuple[()], Vec[ErrorDetail]]:
    should_skip = skip_doctest or lines.iter().any(
        lambda line: bool(CODE_BLOCK_PATTERN.search(line.lstrip()) and "python" in line)
    )
    if should_skip:
        return Ok(())
    return Err(Vec([ErrorDetail(start_line, ErrorKind.NO_DOCTEST)]))


if __name__ == "__main__":
    main()

File: crates/pyochain_build/src/check_docstrings.rs
use comfy_table::{Attribute, Cell, Color, ContentArrangement, Table};
use owo_colors::OwoColorize;
use ruff_python_ast::{Expr, Stmt, statement_visitor, statement_visitor::StatementVisitor};
use ruff_python_parser::parse_module;
use ruff_text_size::TextSize;
use std::{
    error::Error,
    ffi::OsStr,
    fmt, fs, io,
    path::{Path, PathBuf},
};

use crate::paths;

const SKIP_DECORATORS: [&str; 4] = ["overload", "override", "wraps", "property"];
const BLOCK_MARKER: &str = "```";
const RICH_TABLE_STYLE: &str = "││━━┡━╇┩│    ┳┻┏┓┗┛";

#[derive(Clone, Copy)]
enum DiagnosticKind {
    Missing,
    NoDoctest,
    Unclosed,
    ClosedNotOpened,
}

impl fmt::Display for DiagnosticKind {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(match self {
            Self::Missing => "Missing docstring",
            Self::NoDoctest => "No ```python block found in docstring",
            Self::Unclosed => "Unclosed ``` block",
            Self::ClosedNotOpened => "Closing ``` block without opening",
        })
    }
}

struct Diagnostic {
    line_no: usize,
    kind: DiagnosticKind,
}

struct DocstringError {
    file_path: PathBuf,
    function_name: String,
    error_line_no: usize,
    diagnostics: Vec<Diagnostic>,
}

struct FileChecker<'source> {
    file_path: &'source Path,
    source: &'source str,
    errors: Vec<DocstringError>,
}

impl FileChecker<'_> {
    fn check_node(
        &mut self,
        name: &str,
        node_start: TextSize,
        decorators: &[ruff_python_ast::Decorator],
        body: &[Stmt],
        is_function: bool,
    ) {
        if has_skip_decorator(decorators) {
            return;
        }

        let node_line = line_number(self.source, node_start);
        match get_docstring(body) {
            None if is_function => self.errors.push(DocstringError {
                file_path: self.file_path.to_path_buf(),
                function_name: name.to_string(),
                error_line_no: node_line,
                diagnostics: vec![Diagnostic {
                    line_no: node_line,
                    kind: DiagnosticKind::Missing,
                }],
            }),
            None => {}
            Some((content, content_start)) => {
                let start_line = line_number(self.source, content_start);
                let mut diagnostics = check_code_blocks(content, start_line);
                if !has_python_block(content) {
                    diagnostics.insert(
                        0,
                        Diagnostic {
                            line_no: start_line,
                            kind: DiagnosticKind::NoDoctest,
                        },
                    );
                }
                if !diagnostics.is_empty() {
                    let error_line_no = diagnostics.first().expect("non-empty diagnostics").line_no;
                    self.errors.push(DocstringError {
                        file_path: self.file_path.to_path_buf(),
                        function_name: name.to_string(),
                        error_line_no,
                        diagnostics,
                    });
                }
            }
        }
    }
}

impl<'ast> StatementVisitor<'ast> for FileChecker<'_> {
    fn visit_stmt(&mut self, statement: &'ast Stmt) {
        match statement {
            Stmt::FunctionDef(function) => self.check_node(
                function.name.as_str(),
                function.range.start(),
                &function.decorator_list,
                &function.body,
                true,
            ),
            Stmt::ClassDef(class) => self.check_node(
                class.name.as_str(),
                class.range.start(),
                &class.decorator_list,
                &class.body,
                false,
            ),
            _ => {}
        }
        statement_visitor::walk_stmt(self, statement);
    }
}

pub(super) fn run(stubs: &paths::Related) -> Result<(), Box<dyn Error>> {
    let files = stubs
        .iter()
        .filter(|path| path.file_name() != Some(OsStr::new("_types.pyi")))
        .collect::<Vec<_>>();
    anstream::eprintln!("Checking docstrings for properly closed code blocks...");
    anstream::eprintln!("Checking {} pyi files...", files.len());

    let errors = files.iter().try_fold(Vec::new(), |mut errors, path| {
        errors.extend(check_file(path)?);
        Ok::<_, Box<dyn Error>>(errors)
    })?;

    if errors.is_empty() {
        anstream::eprintln!("{}", "[OK] No issues found!".green());
    } else {
        show_errors(&errors, stubs);
        anstream::eprintln!(
            "{}",
            format!("[WARNING] Found {} issue(s)", errors.len())
                .yellow()
                .bold()
        );
    }
    Ok(())
}

fn check_file(file_path: &Path) -> Result<Vec<DocstringError>, Box<dyn Error>> {
    let source = fs::read_to_string(file_path)?;
    let parsed = parse_module(&source).map_err(|error| {
        io::Error::new(
            io::ErrorKind::InvalidData,
            format!("Failed to parse {}: {error}", file_path.display()),
        )
    })?;
    let mut checker = FileChecker {
        file_path,
        source: &source,
        errors: Vec::new(),
    };
    checker.visit_body(parsed.suite());
    Ok(checker.errors)
}

fn show_errors(errors: &[DocstringError], stubs: &paths::Related) {
    let mut table = Table::new();
    table
        .load_preset(RICH_TABLE_STYLE)
        .set_content_arrangement(ContentArrangement::Dynamic)
        .set_header(vec![
            Cell::new("File").add_attribute(Attribute::Bold),
            Cell::new("Function").add_attribute(Attribute::Bold),
            Cell::new("Error").add_attribute(Attribute::Bold),
        ])
        .add_rows(errors.iter().map(|error| {
            let path = format!(
                "{}:{}",
                stubs.make_relative(&error.file_path).0.display(),
                error.error_line_no
            );
            let diagnostics = error
                .diagnostics
                .iter()
                .map(|diagnostic| diagnostic.kind.to_string())
                .collect::<Vec<_>>()
                .join("\n");
            vec![
                Cell::new(path).fg(Color::Cyan),
                Cell::new(&error.function_name).fg(Color::Magenta),
                Cell::new(diagnostics).fg(Color::Red),
            ]
        }));
    anstream::eprintln!("{}", "Issues Found".bold());
    anstream::eprintln!("{table}");
}

fn get_docstring(body: &[Stmt]) -> Option<(&str, TextSize)> {
    let Some(Stmt::Expr(expression)) = body.first() else {
        return None;
    };
    let Expr::StringLiteral(string) = expression.value.as_ref() else {
        return None;
    };
    Some((string.value.to_str(), string.range.start()))
}

fn has_skip_decorator(decorators: &[ruff_python_ast::Decorator]) -> bool {
    decorators
        .iter()
        .any(|decorator| match &decorator.expression {
            Expr::Name(name) => SKIP_DECORATORS.contains(&name.id.as_str()),
            Expr::Attribute(attribute) => SKIP_DECORATORS.contains(&attribute.attr.as_str()),
            _ => false,
        })
}

fn check_code_blocks(docstring: &str, start_line: usize) -> Vec<Diagnostic> {
    let (open_blocks, mut diagnostics) = docstring.lines().enumerate().fold(
        (Vec::new(), Vec::new()),
        |(mut open_blocks, mut diagnostics), (line_num, line)| {
            let stripped = line.trim_start();
            if stripped.starts_with(BLOCK_MARKER) {
                if stripped == BLOCK_MARKER {
                    if open_blocks.pop().is_none() {
                        diagnostics.push(Diagnostic {
                            line_no: start_line + line_num,
                            kind: DiagnosticKind::ClosedNotOpened,
                        });
                    }
                } else {
                    open_blocks.push(line_num + 1);
                }
            }
            (open_blocks, diagnostics)
        },
    );
    diagnostics.extend(open_blocks.into_iter().map(|line_num| Diagnostic {
        line_no: start_line + line_num - 1,
        kind: DiagnosticKind::Unclosed,
    }));
    diagnostics
}

fn has_python_block(docstring: &str) -> bool {
    docstring.lines().any(|line| {
        let stripped = line.trim_start();
        stripped.starts_with(BLOCK_MARKER) && line.contains("python")
    })
}

fn line_number(source: &str, offset: TextSize) -> usize {
    source[..offset.to_usize()].split('\n').count()
}
